cpp/ql/src/semmle/code/cpp/models/implementations/Strset.qll

```diff
@@ -0,0 +1,52 @@
+/**
+ * Provides implementation classes modeling `strset` and various similar
+ * functions. See `semmle.code.cpp.models.Models` for usage information.
+ */
+
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.DataFlow
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+/**
+ * The standard function `strset` and its assorted variants
+ */
+private class StrsetFunction extends ArrayFunction, DataFlowFunction, AliasFunction,
+  SideEffectFunction {
+  StrsetFunction() {
+    hasGlobalName([
+        "strset", "_strset", "_strset_l", "_wcsset", "_wcsset_l", "_mbsset", "_mbsset_l",
+        "_mbsnbset", "_mbsnbset_l", "_strnset", "_strnset_l", "_wcsnset", "_wcsnset_l", "_mbsnset",
+        "_mbsnset_l"
+      ])
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    // flow from the character that overrides the string
+    input.isParameter(1) and
+    (
+      output.isReturnValueDeref()
+      or
+      output.isParameterDeref(1)
+    )
+    or
+    // flow from the input string to the output string
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+
+  override predicate parameterNeverEscapes(int index) { none() }
+
+  override predicate parameterEscapesOnlyViaReturn(int index) { index = 0 }
+
+  override predicate parameterIsAlwaysReturned(int index) { index = 0 }
+
+  override predicate hasOnlySpecificReadSideEffects() { any() }
```

Comment:
Then shouldn't we override `hasSpecificReadSideEffect`?  I'm not entirely sure what this stuff is used for though.

Reply:
Good point. I think you're right: we should implement `hasSpecificReadSideEffect` (and also `hasSpecificWriteSideEffect`).

The way I like to think about `hasSpecificReadSideEffect` is that it's expressing that "the only memory that will be read is the memory specific by `hasSpecificReadSideEffect`". So by implementing `hasSpecificReadSideEffect` as `any()` and `hasSpecificReadSideEffect` as `none()` we're saying "there will be no reads at all". And that's not correct for this set of functions, as `strset` does actually read the memory it's given. I will implement `hasSpecificReadSideEffect` to reflect this.